hypothesis-python/src/hypothesis/searchstrategy/strategies.py

```diff
@@ -624,6 +624,37 @@ def do_draw(self, data):
                 # As long as we consume data, we'll eventually pass or raise.
                 # But if we don't this could be an infinite loop.
                 assume(data.index > start_index)
+
+        # Special-case for SampledFrom strategy.  This is special even among
+        # strategies with a known and small(ish) list of possible elements,
+        # because it implements drawing directly as choice of an index.
+        # That means we can use the "make your own luck" trick explained in
+        # the shrinking guide to find even very rare allowed values.
+        from hypothesis.searchstrategy.lazy import unwrap_strategies
+        from hypothesis.searchstrategy.misc import SampledFromStrategy
+
+        strategy = unwrap_strategies(self.filtered_strategy)
```

Comment:
Not sure if it's worth it, but we could cache this locally (in e.g. `self.__unwrapped_strategy`) to avoid having to look it up in a global map during every single draw.

Reply:
IMO it's not - we only get to this bit if we've failed to draw by rejection sampling, and a function call + lookup is much faster than that.